### PR TITLE
feat: added managed inference integrations for LangChain

### PR DESCRIPTION
**Pull request by justinwilaby**
[W-19066978](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00002IK0yPYAT/view)
# Managed Inference and Agents integration for LangChain
```mermaid
flowchart TB
    %% Heroku-themed, vertical layout
    style Heroku fill:#430098,stroke:#fff,stroke-width:2px,color:#fff
    style InferenceAPI fill:#8c30f5,stroke:#fff,stroke-width:2px,color:#fff
    style MiaChat fill:#b47aff,stroke:#fff,stroke-width:2px,color:#fff
    style LangChain fill:#333333,stroke:#888
    style UserApp fill:#401075,stroke:#fffde7
    style EnvConfig fill:#333333,stroke:#eeeeee,stroke-width:1px

    UserApp["User Applications"]
    MiaChat["MiaChat - (Custom LLM)"]
    LangChain["LangChain Core - (BaseChatModel, Messages, Outputs)"]
    InferenceAPI["Heroku Inference API - /v1/chat/completions"]
    Heroku["Heroku Managed APIs"]
    EnvConfig["Environment Configuration<br/>(INFERENCE_URL, INFERENCE_KEY,<br/>INFERENCE_MODEL_ID)"]

    UserApp -->|"calls"| MiaChat
    MiaChat -->|"inherits"| LangChain
    MiaChat -->|"sends chat completion requests"| InferenceAPI
    InferenceAPI -->|"returns AIMessage"| MiaChat
    MiaChat -->|"returns AIMessage to user code"| UserApp
    InferenceAPI --> Heroku
    EnvConfig -->|"configures"| MiaChat
    EnvConfig -->|"configures"| UserApp
```

## Overview

This PR provides **LangChain integration for the Heroku Managed Inference API** via the `MiaChat` class. It improves documentation, testability, and example usage in the `langchain-heroku` project.

---

## Key Features

- **LangChain integration for Heroku Inference API** via `MiaChat` class
- **Comprehensive docstring examples** including environment setup, basic, advanced, and streaming usage
- **Test harness** in `test_harness.py` with comprehensive test suite
- **Integration tests** in `tests/integration_tests/test_production_endpoint.py`
- **Example usage scripts** in `example_usage.py`
- **Environment variable configuration** with dotenv support
- **Support for all message types** (SystemMessage, HumanMessage, AIMessage, ToolMessage, FunctionMessage)
- **Streaming support** and extended thinking features for Claude Sonnet models

---

## Technical Details

- `MiaChat` class inherits from LangChain's `BaseChatModel`
- Comprehensive test suite includes basic conversation, system messages, multi-turn conversations, tool messages, function messages, streaming, temperature variation, max tokens, stop sequences, and error handling
- Supports both `chat.invoke(messages)` and `chat.stream(messages)`
- Integration tests validate against production endpoints
- Environment configuration via dotenv files or system environment variables

---

### Prerequisites
- Python 3.8+
- Dependencies: `poetry install --with test`
- (Optional) Valid Heroku Inference API credentials

### Commands

```bash
# Manual test harness
python test_harness.py --env .env

# Example usage
python example_usage.py

# Unit tests
poetry run pytest tests/unit_tests/

# Integration tests
poetry run pytest tests/integration_tests/test_production_endpoint.py
```

### Environment Variables
- `INFERENCE_URL`
- `INFERENCE_KEY` 
- `INFERENCE_MODEL_ID`

---

## Impact

- **LangChain integration for Heroku Inference API** via `MiaChat`
- **Improved developer experience** with clearer documentation and examples
- **Comprehensive testing** with manual and automated test suites
- **Production-ready** with integration tests against real endpoints
- **Reduced support burden** with better error handling and debugging

---

## Future Considerations

- Async/streaming test harnesses
- Additional error handling and edge case tests
- CI/CD integration for automated environment-based tests
- Support for additional Heroku API endpoints

### CODE REVIEW

**Review by michaelmalave — COMMENTED** [review]
Just smaller comments to consider:
1. Might be good to include the required environment variables in the root README. Possibly linking to the MIA documentation?
https://devcenter.heroku.com/articles/heroku-inference#model-resource-config-vars

I dont think I see a requirements.txt file that is referenced in the PR description.

File: langchain-heroku/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+[build-system]
```
Comment (hillarysanders) [inline langchain-heroku/pyproject.toml]: Usually these files & folders inside langchain-heroku would be in the top-level directory of a repo - is there a reason they've been put inside langchain-heroku/?

File: langchain-heroku/tests/__init__.py
Comment (hillarysanders) [inline langchain-heroku/tests/__init__.py:1]: Can you add a README in this folder on how to run tests?

File: langchain-heroku/Makefile [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+# Default target executed when no arguments are given to make.
+all: help
+
+# Define a variable for the test file path.
+TEST_FILE ?= tests/unit_tests/
+integration_test integration_tests: TEST_FILE = tests/integration_tests/
+
+
+# unit tests are run with the --disable-socket flag to prevent network calls
+test tests:
```
Comment (hillarysanders) [inline langchain-heroku/Makefile]: I get this:
```
poetry run pytest --disable-socket --allow-unix-socket tests/unit_tests/
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --disable-socket --allow-unix-socket
  inifile: /Users/hillary.sanders/langchain-heroku/langchain-heroku/pyproject.toml
  rootdir: /Users/hillary.sanders/langchain-heroku/langchain-heroku

make: *** [test] Error 4
```

Unless I run `poetry install --with test` --> can that be added to the tests/ README or similar?

File: langchain-heroku/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+`ChatHeroku` class exposes chat models from Heroku.
+
+```python
+from langchain_heroku import ChatHeroku
+
+llm = ChatHeroku()
+llm.invoke("Sing a ballad of LangChain.")
+```
+
+## Embeddings
```
Comment (hillarysanders) [inline langchain-heroku/README.md]: Not implemented yet right? --> comment out?

File: langchain-heroku/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# langchain-heroku
+
+This package contains the LangChain integration with Heroku
+
+## Installation
+
+```bash
+pip install -U langchain-heroku
+```
```
Comment (hillarysanders) [inline langchain-heroku/README.md]: --> Is the idea that this would work once published to pypi?

File: langchain-heroku/docs/chat.ipynb [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    "\n",
+    "- TODO: Fill in table features.\n",
+    "- TODO: Remove JS support link if not relevant, otherwise ensure link is correct.\n",
+    "- TODO: Make sure API reference links are correct.\n",
+    "\n",
+    "| Class | Package | Local | Serializable | [JS support](https://js.langchain.com/docs/integrations/chat/heroku) | Package downloads | Package latest |\n",
+    "| :--- | :--- | :---: | :---: |  :---: | :---: | :---: |\n",
+    "| [ChatHeroku](https://python.langchain.com/api_reference/heroku/chat_models/langchain_heroku.chat_models.ChatHeroku.html) | [langchain-heroku](https://python.langchain.com/api_reference/heroku/) | ✅/❌ | beta/❌ | ✅/❌ | ![PyPI - Downloads](https://img.shields.io/pypi/dm/langchain-heroku?style=flat-square&label=%20) | ![PyPI - Version](https://img.shields.io/pypi/v/langchain-heroku?style=flat-square&label=%20) |\n",
+    "\n",
+    "### Model features\n",
```
Comment (hillarysanders) [inline langchain-heroku/docs/chat.ipynb]: This stuff still needs to be filled out yeah --> saving for a later PR?

File: langchain-heroku/Makefile [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	poetry run pytest --disable-socket --allow-unix-socket $(TEST_FILE)
+
+test_watch:
```
Comment (hillarysanders) [inline langchain-heroku/Makefile]:
```
(base) ~/langchain-heroku/langchain-heroku (venv) (feat/integrations)% make test_watch
poetry run ptw --snapshot-update --now . -- -vv tests/unit_tests/

The option "--snapshot-update" does not exist
make: *** [test_watch] Error 1
```

Am I doing this wrong?

File: langchain-heroku/Makefile [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	poetry run ptw --snapshot-update --now . -- -vv $(TEST_FILE)
+
+# integration tests are run without the --disable-socket flag to allow network calls
+integration_test integration_tests:
```
Comment (hillarysanders) [inline langchain-heroku/Makefile]: Any special setup steps or requirements involved? I ran:
```
export INFERENCE_MODEL_ID=$(heroku config:get -a $APP_NAME INFERENCE_MODEL_ID)
export INFERENCE_KEY=$(heroku config:get -a $APP_NAME INFERENCE_KEY)
export INFERENCE_URL=$(heroku config:get -a $APP_NAME INFERENCE_URL)
```

and then `make integration_test`, & got `9 failed, 2 passed, 28 skipped, 8 warnings in 4.51s`

File: langchain-heroku/langchain_heroku/chat_models.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        result = chat([HumanMessage(content="Hello!")])
+        print(result.generations[0].message.content)
+
+    Streaming usage:
+        chat = ChatHeroku(stream=True)
+        for chunk in chat.stream([HumanMessage(content="Hello!")]):
+            print(chunk.message.content, end="")
+
+    """
+
```
Comment (hillarysanders) [inline langchain-heroku/langchain_heroku/chat_models.py]: Missing `extended_thinking` here, right?

File: langchain-heroku/langchain_heroku/chat_models.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    model_name: Optional[str] = Field(default=None, alias="model")
+    temperature: Optional[float] = None
+    max_tokens: Optional[int] = None
+    timeout: Optional[int] = None
+    stop: Optional[List[str]] = None
+    max_retries: int = 2
+    api_key: Optional[str] = None
+    inference_url: Optional[str] = None
+    tool_schemas: Optional[List[dict]] = None  # For tool calling
```
Comment (hillarysanders) [inline langchain-heroku/langchain_heroku/chat_models.py]: tool_schemas? You mean `tools`? `tool_schemas` isn't a supported MIA parameter - why the name change?

File: langchain-heroku/langchain_heroku/chat_models.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+            print(chunk.message.content, end="")
+
+    """
+
+    model_name: Optional[str] = Field(default=None, alias="model")
+    temperature: Optional[float] = None
+    max_tokens: Optional[int] = None
+    timeout: Optional[int] = None
+    stop: Optional[List[str]] = None
+    max_retries: int = 2
```
Comment (hillarysanders) [inline langchain-heroku/langchain_heroku/chat_models.py]: max_retries is only applicable to the v1/agents/heroku endpoint, which this PR doesn't support, right?

File: langchain-heroku/langchain_heroku/chat_models.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        chat = ChatHeroku(stream=True)
+        for chunk in chat.stream([HumanMessage(content="Hello!")]):
+            print(chunk.message.content, end="")
+
+    """
+
+    model_name: Optional[str] = Field(default=None, alias="model")
+    temperature: Optional[float] = None
+    max_tokens: Optional[int] = None
+    timeout: Optional[int] = None
```
Comment (hillarysanders) [inline langchain-heroku/langchain_heroku/chat_models.py]: May want to decrease the default timeout of 60 to 30 to align w/ heroku's timeout?

File: langchain-heroku/langchain_heroku/chat_models.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+        print(result.generations[0].message.content)
+
+    Streaming usage:
+        chat = ChatHeroku(stream=True)
+        for chunk in chat.stream([HumanMessage(content="Hello!")]):
+            print(chunk.message.content, end="")
+
+    """
+
+    model_name: Optional[str] = Field(default=None, alias="model")
```
Comment (hillarysanders) [inline langchain-heroku/langchain_heroku/chat_models.py]: Called just `model` in the MIA API - why the name change?

File: langchain-heroku/langchain_heroku/chat_models.py [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+    temperature: Optional[float] = None
+    max_tokens: Optional[int] = None
+    timeout: Optional[int] = None
+    stop: Optional[List[str]] = None
+    max_retries: int = 2
+    api_key: Optional[str] = None
+    inference_url: Optional[str] = None
+    tool_schemas: Optional[List[dict]] = None  # For tool calling
+    stream: bool = False
```
Comment (hillarysanders) [inline langchain-heroku/langchain_heroku/chat_models.py]: Missing `tool_choice` parameter?

File: langchain-heroku/langchain_heroku/chat_models.py [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+
```
Comment (hillarysanders) [inline langchain-heroku/langchain_heroku/chat_models.py]: Missing `top_p` parameter?

File: langchain-heroku/langchain_heroku/chat_models.py [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+                        if run_manager:
+                            run_manager.on_llm_new_token(ai_msg.content, chunk=chunk)
+                        yield chunk
+                break
+            except Exception as e:
+                last_exc = e
+        else:
+            raise RuntimeError(f"Heroku Inference API stream call failed after {self.max_retries} retries: {last_exc}")
+
+    # TODO: Implement if ChatHeroku supports async streaming. Otherwise delete.
```
Comment (hillarysanders) [inline langchain-heroku/langchain_heroku/chat_models.py]: async streaming?

File: langchain-heroku/langchain_heroku/chat_models.py [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+import httpx
+
+from langchain_core.callbacks import (
+    CallbackManagerForLLMRun,
+)
+from langchain_core.language_models import BaseChatModel
+from langchain_core.messages import (
+    AIMessage,
+    AIMessageChunk,
+    BaseMessage,
```
Comment (hillarysanders) [inline langchain-heroku/langchain_heroku/chat_models.py]: --> Does this support all the messages types (https://devcenter.heroku.com/articles/heroku-inference-api-v1-chat-completions#messages-array-of-objects) our API supports?

**Review by hillarysanders — DISMISSED** [review]
Awesome to see PRs like this!!

Seems like we've got some MIA v1/chat/completions parameter confusion (https://devcenter.heroku.com/articles/heroku-inference-api-v1-chat-completions) that needs to be fixed up.

I'm also running into issues running the tests, so added some comments & questions there - do they all run successfully for you?

Great start!

Marking request-changes as there are a couple relevant changes that I think should be made, but also this is just a private repo so if it's easier to land and then implement add'l changes, ofc also fine to do that (lmk). Cheers!

File: langchain-heroku/pyproject.toml (hunk 1 above)
Comment (justinwilaby) [reply langchain-heroku/pyproject.toml]: This was scaffolded using the `langchain-cli` and I suspected this wasn't ideal. Glad to see that you seem to agree.  I'll make the update.

File: langchain-heroku/README.md (hunk 3 above)
Comment (justinwilaby) [reply langchain-heroku/README.md]: I'll implement this. Seems like I missed it.

File: langchain-heroku/README.md (hunk 4 above)
Comment (justinwilaby) [reply langchain-heroku/README.md]: Yes. I may need your help here. pypi isn't something I have setup on the Heroku org as of yet.

File: langchain-heroku/langchain_heroku/chat_models.py (hunk 12 above)
Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: Good callout. As I was looking through LangChain's base class, I did see the use of `model_name` for this purpose and kept it with the alias since I wasn't sure if Pydantic would try to serialize `model_name` along with `model`. Maybe this isn't necessary?

Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: So I revisited the sources and it does look like `model` is prioritized when serializing params. I think I'm good to change this.

File: langchain-heroku/langchain_heroku/chat_models.py (hunk 13 above)
Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: Nice catch!

File: langchain-heroku/langchain_heroku/chat_models.py (hunk 15 above)
Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: These are artifacts from the langchain cli scaffolding. Something like a generator equivalent for async streaming I guess? I'm not super familiar with how Python implements generators but in TS it's used quite a lot.

File: langchain-heroku/langchain_heroku/chat_models.py (hunk 16 above)
Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: It should unless I missed something. Some tool fields specific to Heroku were omitted. I hope this is ok.

File: langchain-heroku/Makefile (hunk 2 above)
Comment (justinwilaby) [reply langchain-heroku/Makefile]: This should be fixed. Let me know if you're still running into issues.

File: langchain-heroku/tests/integration_tests/test_chat_models.py
Comment (hillarysanders) [inline langchain-heroku/tests/integration_tests/test_chat_models.py:1]: Why were integration tests deleted?

Comment (hillarysanders) [reply langchain-heroku/tests/integration_tests/test_chat_models.py:1]: Ah, just moved? `git mv` is ideal so the code doesn't look new.

I believe the Makefile needs to be updated.

File: tests/test_chat_models.py [hunk 17]
```diff
@@ -0,0 +1,231 @@
+"""Test chat model integration."""
```
Comment (hillarysanders) [inline tests/test_chat_models.py]: I think it'd be great to add some more advanced tool-calling, multi-message-inference testing here to ensure that the integration works w/ more complicated inference calls (e.g. if you use langchain agents w/ MIA, or if you send MIA back multi-message conversations that include system/tool-call/ etc content)

File: langchain-heroku/langchain_heroku/chat_models.py (hunk 16 above)
Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: Looks like a few were missed. Implemented.

File: langchain-heroku/docs/chat.ipynb (hunk 5 above)
Comment (justinwilaby) [reply langchain-heroku/docs/chat.ipynb]: Done!

File: langchain-heroku/Makefile (hunk 7 above)
Comment (justinwilaby) [reply langchain-heroku/Makefile]: I replaced this with the more idiomatic `dotenv` approach

File: langchain-heroku/langchain_heroku/chat_models.py (hunk 9 above)
Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: Changed

File: langchain-heroku/langchain_heroku/chat_models.py (hunk 8 above)
Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: Yep! it was missing. Nice catch!

File: langchain-heroku/langchain_heroku/chat_models.py (hunk 10 above)
Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: Nice catch! removed.

File: langchain-heroku/langchain_heroku/chat_models.py (hunk 11 above)
Comment (justinwilaby) [reply langchain-heroku/langchain_heroku/chat_models.py]: +1

File: langchain-heroku/Makefile (hunk 6 above)
Comment (justinwilaby) [reply langchain-heroku/Makefile]: Updated.

File: README.md [hunk 18]
```diff
@@ -1,2 +1,109 @@
...
+
+## Installation
+
+```bash
+pip install -U langchain-heroku
+```
+
+And you should configure credentials by setting the following environment variables:
+
+* TODO: fill this out
```
Comment (mandelbro) [inline README.md]: Did you intend to fill this out as part of a later work item?

File: tests/unit_tests/test_chat_models.py [hunk 19]
```diff
@@ -0,0 +1,425 @@
...
+            mock_client = MagicMock()
+            mock_client.post.return_value.json.return_value = mock_response
+            mock_client.post.return_value.raise_for_status.return_value = None
+            mock_client_class.return_value.__enter__.return_value = mock_client
+
+            llm = MiaChat(model="bird-brain-001", temperature=0)
+            messages = [
+                type("Msg", (), {"role": "user", "content": "Say hello"})(),
+            ]
+            result = llm._generate(messages)
```
Comment (mandelbro) [inline tests/unit_tests/test_chat_models.py]: I'm getting the following type error on this line.

```
Argument of type "list[Msg]" cannot be assigned to parameter "messages" of type "List[BaseMessage]" in function "_generate"
  "list[Msg]" is not assignable to "List[BaseMessage]"
    Type parameter "_T@list" is invariant, but "Msg" is not the same as "BaseMessage"
    Consider switching from "list" to "Sequence" which is covariant
```

It looks like the we have it implemented properly in [‎tests/integration_tests/test_production_endpoint.py](https://github.com/heroku/langchain-heroku/pull/1#diff-e88efd234fbcfc3687186e0969e3387e3c1768f495d3a08751ac68d955e8312d)

```
from langchain_core.messages import HumanMessage, SystemMessage
...
messages = [HumanMessage(content="Hello! How are you today?")]
```

Although I think the correct type here is BaseMessage

File: tests/unit_tests/test_chat_models.py [hunk 20]
```diff
@@ -0,0 +1,425 @@
...
+                b'data: {"choices": [{"delta": {"content": " world"}}]}',
+                b"data: [DONE]",
+            ]
+            mock_stream.return_value.__enter__.return_value = mock_response
+
+            llm = MiaChat(model="bird-brain-001", temperature=0, extended_thinking=extended_thinking_config, streaming=True)
+            messages = [type("Msg", (), {"role": "user", "content": "Say hello"})()]
+
+            # Trigger streaming to ensure extended_thinking is included in the payload
+            list(llm._stream(messages))
```
Comment (mandelbro) [inline tests/unit_tests/test_chat_models.py]: Same error here as on line 51

```
Argument of type "list[Msg]" cannot be assigned to parameter "messages" of type "List[BaseMessage]" in function "_stream"
  "list[Msg]" is not assignable to "List[BaseMessage]"
    Type parameter "_T@list" is invariant, but "Msg" is not the same as "BaseMessage"
    Consider switching from "list" to "Sequence" which is covariant
```

File: langchain_heroku/chat_models.py [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+                    resp = client.post(f"{url}/v1/chat/completions", json=payload, headers=headers)
+                    resp.raise_for_status()
+                    data = resp.json()
+                    break
+            except Exception as e:
+                last_exc = e
+        else:
+            raise RuntimeError(f"Heroku Inference API call failed after {max_retries} retries: {last_exc}")
+        ai_msg = self._api_to_ai_message(data)
+        return ChatResult(generations=[ChatGeneration(message=ai_msg)])
```
Comment (mandelbro) [inline langchain_heroku/chat_models.py]: This method is getting pretty long and is doing multiple things, validating config variables, extracting the payload object,  and then performing the actual client request, it could stand to be broken up into several single-purpose methods to make it easier to reason about the code.

File: langchain_heroku/chat_models.py [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+                                    run_manager.on_llm_new_token(content, chunk=chunk)
+                                yield chunk
+                            except json.JSONDecodeError:
+                                # Skip malformed JSON lines
+                                continue
+                break
+            except Exception as e:
+                last_exc = e
+        else:
+            raise RuntimeError(f"Heroku Inference API stream call failed after {max_retries} retries: {last_exc}")
```
Comment (mandelbro) [inline langchain_heroku/chat_models.py]: Same thing as above, there also looks to be a repeated pattern that could be extracted for both long methods.

**Comment by mandelbro** [comment]
General question about the test coverage, I noticed there are no tests for potential errors coming from the `httpx.Client` or `httpx.stream`, but we do have some basic error checking. It would definitely help to add at least one test for error checking in the `MiaChat._generate` and `MiaChat._stream` methods, but I was wondering if we have a plan to add more thorough error handling as well as input and response validation?

**Review by mandelbro — APPROVED** [review]

File: tests/unit_tests/test_chat_models.py (hunk 19 above)
Comment (justinwilaby) [reply tests/unit_tests/test_chat_models.py]: Interesting. My type checker and mypy didn't catch this. 🤔
